Panics if you attempt to commit a file that isn't tracked

To reproduce

```
$ dot init ~/.bashrc
Initialized /home/nicolas/.bashrc as "bashrc"

$ dot commit bashrc
Created /home/nicolas/.dotfile/bashrc/786bc4cd0fdcb72c461b8f5465304e6bbc7c91c8

$ dot commit bashr
panic: runtime error: invalid memory address or nil pointer dereference
```

--- a/cli/checkout.go
+++ b/cli/checkout.go
@@ -8,19 +8,24 @@ import (
 )
 
 type checkoutCommand struct {
-	storage    *file.Storage
+	getStorage func() (*file.Storage, error)
 	fileName   string
 	commitHash string
 }
 
 func (c *checkoutCommand) run(ctx *kingpin.ParseContext) error {
+	_, err := c.getStorage()
+	if err != nil {
+		return err
+	}
+
 	fmt.Printf("TODO: Checkout %#v commitHash: %#v\n", c.fileName, c.commitHash)
 	return nil
 }
 
-func addCheckoutSubCommandToApplication(app *kingpin.Application, storage *file.Storage) {
+func addCheckoutSubCommandToApplication(app *kingpin.Application, gs func() (*file.Storage, error)) {
 	cc := &checkoutCommand{
-		storage: storage,
+		getStorage: gs,
 	}
 	c := app.Command("checkout", "revert a file to a previously committed state").Action(cc.run)
 	c.Arg("file-name", "file to revert changes in").Required().StringVar(&cc.fileName)

--- a/cli/commands.go
+++ b/cli/commands.go
@@ -23,30 +23,49 @@ func getHome() (string, error) {
 	return home, nil
 }
 
+// Returns a function that initializes dotfile storage.
+// The result function must be ran at the time of a command being run so that
+// the user can override default storage configuration with --storage-dir or --storage-name.
+func getStorageClosure(home string, dir, name *string) func() (*file.Storage, error) {
+	return func() (*file.Storage, error) {
+		storage := &file.Storage{}
+
+		if err := storage.Setup(home, *dir, *name); err != nil {
+			return nil, errors.Wrap(err, "failed to setup dotfile storage")
+		}
+		return storage, nil
+	}
+}
+
 // AddCommandsToApplication initializes the cli.
 func AddCommandsToApplication(app *kingpin.Application) error {
+	var (
+		storageDirectory string
+		storageName      string
+	)
+
 	home, err := getHome()
 	if err != nil {
 		return err
 	}
-	storage := &file.Storage{
-		Home: home,
-	}
+
+	gs := getStorageClosure(home, &storageDirectory, &storageName)
 
 	app.Flag("storage-dir", "The directory where version control storage is stored").
-		Default(fmt.Sprintf("%s/%s", storage.Home, defaultStorageDir)).
-		StringVar(&storage.Dir)
+		Default(fmt.Sprintf("%s/%s", home, defaultStorageDir)).
+		StringVar(&storageDirectory)
 	app.Flag("storage-name", "The main json file that tracks checked in files").
 		Default(defaultStorageName).
-		StringVar(&storage.Name)
+		StringVar(&storageName)
 
-	addInitSubCommandToApplication(app, storage)
-	addEditSubCommandToApplication(app, storage)
-	addDiffSubCommandToApplication(app, storage)
-	addLogSubCommandToApplication(app, storage)
-	addCheckoutSubCommandToApplication(app, storage)
-	addCommitSubCommandToApplication(app, storage)
-	addPushSubCommandToApplication(app, storage)
-	addPullSubCommandToApplication(app, storage)
+	addInitSubCommandToApplication(app, gs)
+	addEditSubCommandToApplication(app, gs)
+	addDiffSubCommandToApplication(app, gs)
+	addLogSubCommandToApplication(app, gs)
+	addCheckoutSubCommandToApplication(app, gs)
+	addCommitSubCommandToApplication(app, gs)
+	addPushSubCommandToApplication(app, gs)
+	addPullSubCommandToApplication(app, gs)
+
 	return nil
 }

--- a/cli/commit.go
+++ b/cli/commit.go
@@ -11,23 +11,27 @@ import (
 const defaultMessageTimestampDisplayFormat = "January 02, 2006 3:04 PM -0700"
 
 type commitCommand struct {
-	storage       *file.Storage
+	getStorage    func() (*file.Storage, error)
 	fileName      string
 	commitMessage string
 }
 
 func (c *commitCommand) run(ctx *kingpin.ParseContext) error {
-	_, err := file.Commit(c.storage, c.fileName, c.commitMessage)
+	s, err := c.getStorage()
 	if err != nil {
+		return err
+	}
+
+	if _, err := file.Commit(s, c.fileName, c.commitMessage); err != nil {
 		return errors.Wrapf(err, "failed to commit %#v", c.fileName)
 	}
 
 	return nil
 }
 
-func addCommitSubCommandToApplication(app *kingpin.Application, storage *file.Storage) {
+func addCommitSubCommandToApplication(app *kingpin.Application, gs func() (*file.Storage, error)) {
 	cc := &commitCommand{
-		storage: storage,
+		getStorage: gs,
 	}
 	c := app.Command("commit", "commit file to working tree").Action(cc.run)
 	c.Arg("file-name", "the file to track").Required().StringVar(&cc.fileName)

--- a/cli/commit_test.go
+++ b/cli/commit_test.go
@@ -1,0 +1,27 @@
+package cli
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCommit(t *testing.T) {
+	clearTestStorage()
+	initTestFile(t)
+
+	commitCommand := &commitCommand{
+		getStorage:    getTestStorageClosure(),
+		commitMessage: "test commit",
+	}
+
+	t.Run("returns error when file is not tracked", func(t *testing.T) {
+		commitCommand.fileName = notTrackedFile
+		assert.Error(t, commitCommand.run(nil))
+	})
+
+	t.Run("ok", func(t *testing.T) {
+		commitCommand.fileName = trackedFileAlias
+		assert.NoError(t, commitCommand.run(nil))
+	})
+}

--- a/cli/diff.go
+++ b/cli/diff.go
@@ -8,18 +8,23 @@ import (
 )
 
 type diffCommand struct {
-	storage  *file.Storage
-	fileName string
+	getStorage func() (*file.Storage, error)
+	fileName   string
 }
 
 func (d *diffCommand) run(ctx *kingpin.ParseContext) error {
+	_, err := d.getStorage()
+	if err != nil {
+		return err
+	}
+
 	fmt.Printf("TODO: Diff %#v\n", d.fileName)
 	return nil
 }
 
-func addDiffSubCommandToApplication(app *kingpin.Application, storage *file.Storage) {
+func addDiffSubCommandToApplication(app *kingpin.Application, gs func() (*file.Storage, error)) {
 	dc := &diffCommand{
-		storage: storage,
+		getStorage: gs,
 	}
 	c := app.Command("diff", "check changes to tracked file").Action(dc.run)
 	c.Arg("file-name", "file to check changes in").Required().StringVar(&dc.fileName)

--- a/cli/edit.go
+++ b/cli/edit.go
@@ -10,8 +10,8 @@ import (
 )
 
 type editCommand struct {
-	storage  *file.Storage
-	fileName string
+	getStorage func() (*file.Storage, error)
+	fileName   string
 }
 
 var (
@@ -26,7 +26,12 @@ func (e *editCommand) run(ctx *kingpin.ParseContext) error {
 		return ErrEditorEnvVarNotSet
 	}
 
-	path, err := file.GetPath(e.storage, e.fileName)
+	s, err := e.getStorage()
+	if err != nil {
+		return err
+	}
+
+	path, err := file.GetPath(s, e.fileName)
 	if err != nil {
 		return errors.Wrapf(err, "error getting path for filename: %#v", e.fileName)
 	}
@@ -38,9 +43,9 @@ func (e *editCommand) run(ctx *kingpin.ParseContext) error {
 	return cmd.Run()
 }
 
-func addEditSubCommandToApplication(app *kingpin.Application, storage *file.Storage) {
+func addEditSubCommandToApplication(app *kingpin.Application, gs func() (*file.Storage, error)) {
 	ec := &editCommand{
-		storage: storage,
+		getStorage: gs,
 	}
 	c := app.Command("edit", "open a tracked file in $EDITOR").Action(ec.run)
 	c.Arg("file-name", "the file to edit").Required().StringVar(&ec.fileName)

--- a/cli/edit_test.go
+++ b/cli/edit_test.go
@@ -33,8 +33,8 @@ func TestEditCommandLaunchesEditor(t *testing.T) {
 	os.Setenv("EDITOR", arbitraryEditor)
 
 	editCommand := &editCommand{
-		fileName: testAlias,
-		storage:  getTestStorage(),
+		fileName:   trackedFileAlias,
+		getStorage: getTestStorageClosure(),
 	}
 
 	clearTestStorage()
@@ -56,7 +56,7 @@ func TestErrorIfEditorNotSet(t *testing.T) {
 	os.Unsetenv("EDITOR")
 
 	command := &editCommand{
-		fileName: testAlias,
+		fileName: trackedFileAlias,
 	}
 	err := command.run(nil)
 	assert.Equal(t, ErrEditorEnvVarNotSet, err)
@@ -66,6 +66,6 @@ func TestEditHelperProcess(t *testing.T) {
 	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
 		return
 	}
-	assert.Equal(t, testAlias, os.Args[1])
+	assert.Equal(t, trackedFileAlias, os.Args[1])
 	os.Exit(0)
 }

--- a/cli/helpers_test.go
+++ b/cli/helpers_test.go
@@ -10,24 +10,24 @@ import (
 )
 
 const (
-	arbitraryFile = "helpers_test.go"
-	testAlias     = "helpers_test"
-	testDir       = "testdata/"
+	nonExistantFile  = "file_does_not_exist"
+	notTrackedFile   = "/dev/null"
+	trackedFile      = "helpers_test.go"
+	trackedFileAlias = "helpers_test"
+	testDir          = "testdata/"
 )
 
-func getTestStorage() *file.Storage {
+func getTestStorageClosure() func() (*file.Storage, error) {
 	home, _ := getHome()
-	return &file.Storage{
-		Home: home,
-		Dir:  testDir,
-		Name: defaultStorageName,
-	}
+	dir := testDir
+	name := defaultStorageName
+	return getStorageClosure(home, &dir, &name)
 }
 
 func initTestFile(t *testing.T) {
 	initCommand := &initCommand{
-		fileName: arbitraryFile,
-		storage:  getTestStorage(),
+		fileName:   trackedFile,
+		getStorage: getTestStorageClosure(),
 	}
 	err := initCommand.run(nil)
 	assert.NoError(t, err)

--- a/cli/init.go
+++ b/cli/init.go
@@ -9,13 +9,18 @@ import (
 )
 
 type initCommand struct {
-	storage  *file.Storage
-	fileName string
-	altName  string
+	getStorage func() (*file.Storage, error)
+	fileName   string
+	altName    string
 }
 
 func (ic *initCommand) run(ctx *kingpin.ParseContext) error {
-	alias, err := file.Init(ic.storage, ic.fileName, ic.altName)
+	s, err := ic.getStorage()
+	if err != nil {
+		return err
+	}
+
+	alias, err := file.Init(s, ic.fileName, ic.altName)
 	if err != nil {
 		return errors.Wrapf(err, "failed to initialize %#v", ic.fileName)
 	}
@@ -24,9 +29,9 @@ func (ic *initCommand) run(ctx *kingpin.ParseContext) error {
 	return nil
 }
 
-func addInitSubCommandToApplication(app *kingpin.Application, storage *file.Storage) {
+func addInitSubCommandToApplication(app *kingpin.Application, gs func() (*file.Storage, error)) {
 	ic := &initCommand{
-		storage: storage,
+		getStorage: gs,
 	}
 	p := app.Command("init", "begin tracking a file").Action(ic.run)
 	p.Arg("file-name", "the file to track").Required().StringVar(&ic.fileName)

--- a/cli/init_test.go
+++ b/cli/init_test.go
@@ -6,13 +6,11 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-const nonExistantFile = "this does not exist"
-
 func TestInit(t *testing.T) {
 	clearTestStorage()
 
 	initCommand := &initCommand{
-		storage: getTestStorage(),
+		getStorage: getTestStorageClosure(),
 	}
 
 	t.Run("returns error when file does not exist", func(t *testing.T) {
@@ -21,7 +19,7 @@ func TestInit(t *testing.T) {
 	})
 
 	t.Run("no error when file exists", func(t *testing.T) {
-		initCommand.fileName = arbitraryFile
+		initCommand.fileName = trackedFile
 		assert.NoError(t, initCommand.run(nil))
 	})
 }

--- a/cli/log.go
+++ b/cli/log.go
@@ -8,18 +8,23 @@ import (
 )
 
 type logCommand struct {
-	storage  *file.Storage
-	fileName string
+	getStorage func() (*file.Storage, error)
+	fileName   string
 }
 
 func (l *logCommand) run(ctx *kingpin.ParseContext) error {
+	_, err := l.getStorage()
+	if err != nil {
+		return err
+	}
+
 	fmt.Printf("TODO: Log %#v\n", l.fileName)
 	return nil
 }
 
-func addLogSubCommandToApplication(app *kingpin.Application, storage *file.Storage) {
+func addLogSubCommandToApplication(app *kingpin.Application, gs func() (*file.Storage, error)) {
 	lc := &logCommand{
-		storage: storage,
+		getStorage: gs,
 	}
 	c := app.Command("log", "shows revision history with commit hashes for a tracked file").Action(lc.run)
 	c.Arg("file-name", "tracked file to show history for").Required().StringVar(&lc.fileName)

--- a/cli/pull.go
+++ b/cli/pull.go
@@ -9,12 +9,17 @@ import (
 )
 
 type pullCommand struct {
-	storage  *file.Storage
-	fileName string
-	pullAll  bool
+	getStorage func() (*file.Storage, error)
+	fileName   string
+	pullAll    bool
 }
 
 func (pc *pullCommand) run(ctx *kingpin.ParseContext) error {
+	_, err := pc.getStorage()
+	if err != nil {
+		return err
+	}
+
 	if pc.pullAll {
 		fmt.Println("TODO: Pull all")
 	} else if pc.fileName != "" {
@@ -25,9 +30,9 @@ func (pc *pullCommand) run(ctx *kingpin.ParseContext) error {
 	return nil
 }
 
-func addPullSubCommandToApplication(app *kingpin.Application, storage *file.Storage) {
+func addPullSubCommandToApplication(app *kingpin.Application, gs func() (*file.Storage, error)) {
 	pc := &pullCommand{
-		storage: storage,
+		getStorage: gs,
 	}
 	p := app.Command("pull", "pull changes from central service").Action(pc.run)
 	p.Arg("file-name", "the file to pull").StringVar(&pc.fileName)

--- a/cli/push.go
+++ b/cli/push.go
@@ -8,18 +8,23 @@ import (
 )
 
 type pushCommand struct {
-	storage  *file.Storage
-	fileName string
+	getStorage func() (*file.Storage, error)
+	fileName   string
 }
 
 func (pc *pushCommand) run(ctx *kingpin.ParseContext) error {
+	_, err := pc.getStorage()
+	if err != nil {
+		return err
+	}
+
 	fmt.Printf("TODO: Push %#v", pc.fileName)
 	return nil
 }
 
-func addPushSubCommandToApplication(app *kingpin.Application, storage *file.Storage) {
+func addPushSubCommandToApplication(app *kingpin.Application, gs func() (*file.Storage, error)) {
 	pc := &pushCommand{
-		storage: storage,
+		getStorage: gs,
 	}
 	p := app.Command("push", "push committed changes to central service").Action(pc.run)
 	p.Arg("file-name", "the file to push").Required().StringVar(&pc.fileName)

--- a/file/files.go
+++ b/file/files.go
@@ -19,14 +19,14 @@ var pathToAliasRegex = regexp.MustCompile(`(\w+)(\.\w+)?$`)
 
 // Init sets up a file for dotfile to track.
 // Returns the alias for the newly tracked file.
-func Init(d *Storage, filePath, altName string) (string, error) {
+func Init(s *Storage, filePath, altName string) (string, error) {
 	var (
 		alias string
 		err   error
 	)
 
 	if _, err = os.Stat(filePath); os.IsNotExist(err) {
-		return "", fmt.Errorf("\"%#v\" not found", filePath)
+		return "", fmt.Errorf("%#v not found", filePath)
 	}
 
 	// Get the full path so that it can later turn it into a relative path.
@@ -43,14 +43,10 @@ func Init(d *Storage, filePath, altName string) (string, error) {
 		}
 	}
 
-	if err = d.setup(); err != nil {
-		return "", err
-	}
-
 	// Replace the full path with a relative path.
-	relativePath := strings.Replace(fullPath, d.Home, "~", 1)
+	relativePath := strings.Replace(fullPath, s.GetHome(), "~", 1)
 
-	if err = d.save(alias, &trackedFile{
+	if err = s.save(alias, &trackedFile{
 		Path: relativePath,
 	}); err != nil {
 		return "", err
@@ -59,13 +55,13 @@ func Init(d *Storage, filePath, altName string) (string, error) {
 }
 
 // Commit hashes and saves the current state of a tracked file.
-func Commit(d *Storage, alias, message string) (string, error) {
-	file, err := d.getTrackedFile(alias)
+func Commit(s *Storage, alias, message string) (string, error) {
+	file, err := s.getTrackedFile(alias)
 	if err != nil {
 		return "", err
 	}
 
-	path := file.getFullPath(d.Home)
+	path := file.getFullPath(s.GetHome())
 	f, err := os.Open(path)
 	if err != nil {
 		return "", errors.Wrapf(err, "failed to open %s", path)
@@ -93,17 +89,17 @@ func Commit(d *Storage, alias, message string) (string, error) {
 
 	file.Commits = append(file.Commits, c)
 
-	return hash, d.saveCommit(c, alias, file, compressed.Bytes())
+	return hash, s.saveCommit(c, alias, file, compressed.Bytes())
 }
 
 // GetPath gets the full path for a tracked file.
-func GetPath(d *Storage, alias string) (string, error) {
-	file, err := d.getTrackedFile(alias)
+func GetPath(s *Storage, alias string) (string, error) {
+	file, err := s.getTrackedFile(alias)
 	if err != nil {
 		return "", err
 	}
 
-	return file.getFullPath(d.Home), nil
+	return file.getFullPath(s.GetHome()), nil
 }
 
 // Creates a alias from the path of the file.

--- a/file/helpers_test.go
+++ b/file/helpers_test.go
@@ -1,0 +1,32 @@
+package file
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	testFile  = "storage_test.go"
+	testAlias = "storage_test"
+	testDir   = "testdata/"
+	testName  = "files.json"
+)
+
+func clearTestStorage() {
+	os.Remove(fmt.Sprintf("%s%s", testDir, testName))
+}
+
+func getTestStorage() *Storage {
+	home, _ := os.UserHomeDir()
+	s := &Storage{}
+	s.Setup(home, testDir, testName)
+	return s
+}
+
+func initTestFile(t *testing.T, s *Storage) {
+	_, err := Init(s, testFile, testAlias)
+	assert.NoError(t, err)
+}

--- a/file/storage.go
+++ b/file/storage.go
@@ -35,22 +35,55 @@ import (
 // Storage provides methods for reading and writing json data and compressed commit bytes.
 // All exported fields should be set.
 type Storage struct {
-	Home string // The path to the users home directory.
-	Dir  string // The path to the folder where data will be stored.
-	Name string // The name of the json file.
+	home string // The path to the users home directory.
+	dir  string // The path to the folder where data will be stored.
+	name string // The name of the json file.
 
 	path  string
 	files map[string]*trackedFile
 }
 
+// Sets up the storage directory if it hasn't been done yet and pulls its data.
+func (s *Storage) Setup(home, dir, name string) error {
+	if home == "" {
+		return errors.New("home cannot be empty  ")
+	}
+	if dir == "" {
+		return errors.New("dir cannot be empty")
+	}
+	if name == "" {
+		return errors.New("name cannot be empty")
+	}
+
+	s.home = home
+	s.dir = dir
+	s.name = name
+
+	if s.dir[len(s.dir)-1] != '/' {
+		s.dir += "/"
+	}
+
+	s.path = fmt.Sprintf("%s%s", s.dir, s.name)
+
+	if err := createIfNotExist(s.dir, s.path); err != nil {
+		return err
+	}
+
+	return s.get()
+}
+
+func (s *Storage) GetHome() string {
+	return s.home
+}
+
 func (s *Storage) saveCommit(c *commit, alias string, t *trackedFile, bytes []byte) error {
 	// Create the directory for the files commits if it doesn't exist
-	commitDir := fmt.Sprintf("%s%s", s.Dir, alias)
+	commitdir := fmt.Sprintf("%s%s", s.dir, alias)
 
 	// The name of the file will be the hash
-	commitPath := fmt.Sprintf("%s/%s", commitDir, c.Hash)
+	commitPath := fmt.Sprintf("%s/%s", commitdir, c.Hash)
 
-	if err := createIfNotExist(commitDir, commitPath); err != nil {
+	if err := createIfNotExist(commitdir, commitPath); err != nil {
 		return err
 	}
 
@@ -63,10 +96,6 @@ func (s *Storage) saveCommit(c *commit, alias string, t *trackedFile, bytes []by
 
 // Reads the json and makes the tracked file map.
 func (s *Storage) get() error {
-	if err := s.setPath(); err != nil {
-		return err
-	}
-
 	s.files = make(map[string]*trackedFile)
 
 	f, err := os.Open(s.path)
@@ -99,7 +128,7 @@ func (s *Storage) getTrackedFile(alias string) (*trackedFile, error) {
 
 	t, ok := s.files[alias]
 	if !ok {
-		errors.New("file not tracked, use 'dot init <file>' first")
+		return nil, fmt.Errorf("%#v not tracked, use 'dot init <file>' first", alias)
 	}
 	return t, nil
 }
@@ -113,43 +142,7 @@ func (s *Storage) save(alias string, t *trackedFile) error {
 		return errors.Wrapf(err, "failed to marshal %s", s.path)
 	}
 
-	if err := s.setPath(); err != nil {
-		return err
-	}
-
 	return ioutil.WriteFile(s.path, json, 0644)
-}
-
-// Sets up the storage directory if it hasn't been done yet and pulls its data.
-func (s *Storage) setup() error {
-	if err := s.setPath(); err != nil {
-		return err
-	}
-
-	if err := createIfNotExist(s.Dir, s.path); err != nil {
-		return err
-	}
-
-	return s.get()
-}
-
-func (s *Storage) setPath() error {
-	if s.Home == "" {
-		return errors.New("home not set")
-	}
-	if s.Dir == "" {
-		return errors.New("dir not set")
-	}
-	if s.Name == "" {
-		return errors.New("name not set")
-	}
-
-	if s.Dir[len(s.Dir)-1] != '/' {
-		s.Dir += "/"
-	}
-
-	s.path = fmt.Sprintf("%s%s", s.Dir, s.Name)
-	return nil
 }
 
 // Creates a directory and a file.

--- a/file/storage_test.go
+++ b/file/storage_test.go
@@ -1,0 +1,44 @@
+package file
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestInit(t *testing.T) {
+	s := &Storage{}
+
+	t.Run("error when home is empty", func(t *testing.T) {
+		assert.Error(t, s.Setup("", testDir, testName))
+	})
+
+	t.Run("error when dir is empty", func(t *testing.T) {
+		assert.Error(t, s.Setup("home", "", testName))
+	})
+
+	t.Run("error when name is empty", func(t *testing.T) {
+		assert.Error(t, s.Setup("home", testDir, ""))
+	})
+
+}
+
+func TestGetTrackedFile(t *testing.T) {
+	clearTestStorage()
+	s := getTestStorage()
+
+	t.Run("returns error when file is not tracked", func(t *testing.T) {
+		f, err := s.getTrackedFile(testAlias)
+		fmt.Println(err)
+		assert.Error(t, err)
+		assert.Nil(t, f)
+	})
+
+	t.Run("ok when file is tracked", func(t *testing.T) {
+		initTestFile(t, s)
+		f, err := s.getTrackedFile(testAlias)
+		assert.NoError(t, err)
+		assert.NotNil(t, f)
+	})
+}

